lib/libmbedtls/core/rsa.c

```diff
@@ -0,0 +1,653 @@
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA224:
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA224:
+	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA224:
+	case TEE_ALG_SHA224:
+	case TEE_ALG_DSA_SHA224:
+	case TEE_ALG_HMAC_SHA224:
+		return MBEDTLS_MD_SHA224;
+#endif
+#if defined(CFG_CRYPTO_SHA256)
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA256:
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256:
+	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA256:
+	case TEE_ALG_SHA256:
+	case TEE_ALG_DSA_SHA256:
+	case TEE_ALG_HMAC_SHA256:
+		return MBEDTLS_MD_SHA256;
+#endif
+#if defined(CFG_CRYPTO_SHA384)
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA384:
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA384:
+	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA384:
+	case TEE_ALG_SHA384:
+	case TEE_ALG_HMAC_SHA384:
+		return MBEDTLS_MD_SHA384;
+#endif
+#if defined(CFG_CRYPTO_SHA512)
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA512:
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA512:
+	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA512:
+	case TEE_ALG_SHA512:
+	case TEE_ALG_HMAC_SHA512:
+		return MBEDTLS_MD_SHA512;
+#endif
+	default:
+		return MBEDTLS_MD_NONE;
+	}
+}
+
+static void rsa_init_from_key_pair(mbedtls_rsa_context *rsa,
+				struct rsa_keypair *key)
+{
+	mbedtls_rsa_init(rsa, 0, 0);
+
+	rsa->E = *(mbedtls_mpi *)key->e;
+	rsa->N = *(mbedtls_mpi *)key->n;
+	rsa->D = *(mbedtls_mpi *)key->d;
+	if (key->p && crypto_bignum_num_bytes(key->p)) {
+		rsa->P = *(mbedtls_mpi *)key->p;
+		rsa->Q = *(mbedtls_mpi *)key->q;
+		rsa->QP = *(mbedtls_mpi *)key->qp;
+		rsa->DP = *(mbedtls_mpi *)key->dp;
+		rsa->DQ = *(mbedtls_mpi *)key->dq;
+	}
+	rsa->len = mbedtls_mpi_size(&rsa->N);
+}
+
+static void mbd_rsa_free(mbedtls_rsa_context *rsa)
+{
+	/* Reset mpi to skip freeing here, those mpis will be freed with key */
+	mbedtls_mpi_init(&rsa->E);
+	mbedtls_mpi_init(&rsa->N);
+	mbedtls_mpi_init(&rsa->D);
+	if (mbedtls_mpi_size(&rsa->P)) {
+		mbedtls_mpi_init(&rsa->P);
+		mbedtls_mpi_init(&rsa->Q);
+		mbedtls_mpi_init(&rsa->QP);
+		mbedtls_mpi_init(&rsa->DP);
+		mbedtls_mpi_init(&rsa->DQ);
+	}
+	mbedtls_rsa_free(rsa);
+}
+
+TEE_Result crypto_acipher_alloc_rsa_keypair(struct rsa_keypair *s,
+					    size_t key_size_bits)
+{
+	memset(s, 0, sizeof(*s));
+	s->e = crypto_bignum_allocate(key_size_bits);
+	if (!s->e)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	s->d = crypto_bignum_allocate(key_size_bits);
+	if (!s->d)
+		goto err;
+	s->n = crypto_bignum_allocate(key_size_bits);
+	if (!s->n)
+		goto err;
+	s->p = crypto_bignum_allocate(key_size_bits);
+	if (!s->p)
+		goto err;
+	s->q = crypto_bignum_allocate(key_size_bits);
+	if (!s->q)
+		goto err;
+	s->qp = crypto_bignum_allocate(key_size_bits);
+	if (!s->qp)
+		goto err;
+	s->dp = crypto_bignum_allocate(key_size_bits);
+	if (!s->dp)
+		goto err;
+	s->dq = crypto_bignum_allocate(key_size_bits);
+	if (!s->dq)
+		goto err;
+
+	return TEE_SUCCESS;
+err:
+	crypto_bignum_free(s->e);
+	crypto_bignum_free(s->d);
+	crypto_bignum_free(s->n);
+	crypto_bignum_free(s->p);
+	crypto_bignum_free(s->q);
+	crypto_bignum_free(s->qp);
```

Comment:
sorry: missing `	crypto_bignum_free(s->dp);`

Reply:
`s->dq` doesn't need freeing here. Either it was allocated successfully and we return OK or if it fails there's nothing to free.

Reply:
yes, sorry.